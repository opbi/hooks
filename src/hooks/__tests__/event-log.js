import eventLogger from '../event-logger';

const logger = {
  info: jest.fn(),
  error: jest.fn(),
};

describe('eventLogger', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('log the success event with param when enabled', async () => {
    const original = async ({ text }) => text;
    const decorated = eventLogger({ logParam: true })(original);
    const result = await decorated({ text: 'yes' }, {}, { logger });

    expect(result).toBe('yes');
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it('log the success event with result when enabled', async () => {
    const original = async () => 'yes';
    const decorated = eventLogger({ logResult: true })(original);
    const result = await decorated({}, {}, { logger });

    expect(result).toBe('yes');
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it('log chained events from upper level function call', async () => {
    const original = () => {};
    const upper = (p, m, c) => {
      return eventLogger()(original)(p, m, c);
    };
    await eventLogger()(upper)({}, {}, { logger });
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it('log the error event when input function fails', async () => {
    const original = () => {
      const e = { message: 'error' };
      throw e;
    };
    const decorated = eventLogger()(original);

    try {
      await decorated({}, {}, { logger });
    } catch (e) {
      expect(e.message).toBe('error');
      expect(logger.info.mock.calls).toMatchSnapshot();
      expect(logger.error.mock.calls).toMatchSnapshot();
    }
  });

  it('log the parsed error when errorParse set', async () => {
    const original = () => {
      const e = new Error('error');
      throw e;
    };
    const decorated = eventLogger({
      errorParser: e => ({
        message: e.message,
        stack: e.stack.substring(0, 20),
      }),
    })(original);

    try {
      await decorated({}, {}, { logger });
    } catch (e) {
      expect(e.message).toBe('error');
      expect(logger.info.mock.calls).toMatchSnapshot();
      expect(logger.error.mock.calls).toMatchSnapshot();
    }
  });
});
