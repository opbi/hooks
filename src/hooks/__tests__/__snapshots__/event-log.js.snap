// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventLogger log chained events from upper level function call 1`] = `
Array [
  Array [
    Object {
      "event": "upper.original",
    },
    "upper.original",
  ],
  Array [
    Object {
      "event": "upper",
    },
    "upper",
  ],
]
`;

exports[`eventLogger log the error event when input function fails 1`] = `Array []`;

exports[`eventLogger log the error event when input function fails 2`] = `
Array [
  Array [
    Object {
      "error": Object {
        "message": "error",
      },
      "event": "original",
    },
    "error",
  ],
]
`;

exports[`eventLogger log the parsed error when errorParse set 1`] = `Array []`;

exports[`eventLogger log the parsed error when errorParse set 2`] = `
Array [
  Array [
    Object {
      "error": Object {
        "message": "error",
        "stack": "Error: error
    at ",
      },
      "event": "original",
    },
    "error",
  ],
]
`;

exports[`eventLogger log the success event with param when enabled 1`] = `
Array [
  Array [
    Object {
      "event": "original",
      "param": Object {
        "text": "yes",
      },
    },
    "original",
  ],
]
`;

exports[`eventLogger log the success event with result when enabled 1`] = `
Array [
  Array [
    Object {
      "event": "original",
      "result": "yes",
    },
    "original",
  ],
]
`;
